ddtrace/profiling/collector/_memalloc.c

```diff
@@ -42,113 +42,81 @@ static PyObject* object_string = NULL;
-    // there's nothing to do.
-    if (!memlock_trylock(&g_memalloc_lock)) {
+    /* Return if we can't take the guard, since we're not going to be able to sample */
+    if (!memalloc_take_guard()) {
         return;
     }
 
     /* Determine if we can capture or if we need to sample */
-    if (global_alloc_tracker->allocs.count < ctx->max_events) {
+    bool should_sample = false;
+    if (alloc_count < ctx->max_events) {
         /* Buffer is not full, fill it */
-        traceback_t* tb = memalloc_get_traceback(ctx->max_nframe, ptr, size, ctx->domain);
-        if (tb) {
-            traceback_array_append(&global_alloc_tracker->allocs, tb);
-        }
+        should_sample = true;
     } else {
         /* Sampling mode using a reservoir sampling algorithm: replace a random
-         * traceback with this one */
+         * traceback with this one
+         * NB: this just decides whether we sample. See comment below;
+         * we will probably have to recompute the index to replace */
         uint64_t r = random_range(alloc_count);
+        should_sample = r < ctx->max_events;
+    }
 
-        // In addition to event size, need to check that the tab is in a good state
-        if (r < ctx->max_events && global_alloc_tracker->allocs.tab != NULL) {
-            /* Replace a random traceback with this one */
-            traceback_t* tb = memalloc_get_traceback(ctx->max_nframe, ptr, size, ctx->domain);
+    if (!should_sample) {
+        goto done;
+    }
 
-            // Need to check not only that the tb returned
-            if (tb) {
-                traceback_free(global_alloc_tracker->allocs.tab[r]);
-                global_alloc_tracker->allocs.tab[r] = tb;
-            }
-        }
+    traceback_t* tb = memalloc_get_traceback(ctx->max_nframe, ptr, size, ctx->domain);
+    if (!tb) {
+        goto done;
     }
 
-    memlock_unlock(&g_memalloc_lock);
+    if (!global_alloc_tracker) {
+        /* If getting a traceback lead to a GIL release, it is possible
+         * that the profiler was stopped and the traceback array was freed.
+         * Don't add the sample.
+         */
+        traceback_free(tb);
+        goto done;
+    }
+
+    /* The GIL may have been released while collecting a traceback above.
+     * If so, the aggregation/uploading thread may have rotated the tracker.
+     * We need to recompute the index where the sample will go.
+     */
+    if (global_alloc_tracker->allocs.count < ctx->max_events) {
+        traceback_array_append(&global_alloc_tracker->allocs, tb);
+    } else {
+        uint64_t r = random_range(ctx->max_events);
+        traceback_t* old = global_alloc_tracker->allocs.tab[r];
+        global_alloc_tracker->allocs.tab[r] = tb;
+        /* Free the old traceback only after modifying the table, in case freeing
+         * releases the GIL */
+        traceback_free(old);
+    }
+
+done:
     memalloc_yield_guard();
```

Comment:
Instead of using goto, we can wrap the critical sections inside an `_inner` function, and the outer function is responsible for unconditionally entering and exiting the critical section